classes/class-kco-gateway.php

```diff
@@ -86,6 +86,86 @@ public function __construct() {
 
 			add_filter( 'woocommerce_order_needs_payment', array( $this, 'maybe_change_needs_payment' ), 999, 3 );
 			add_filter( 'kco_wc_api_request_args', array( $this, 'maybe_remove_kco_epm' ), 9999 );
+
+			// Prevent the Woo validation from proceeding if there is a discrepancy between Woo and Klarna.
+			add_action( 'woocommerce_after_checkout_validation', array( $this, 'validate_checkout' ), 10, 2 );
+		}
+
+		/**
+		 * Validate the data of the checkout fields matches the Klarna order.
+		 *
+		 * @param array     $data An array of posted data.
+		 * @param WP_Errors $errors Validation errors.
+		 * @return void
+		 */
+		public function validate_checkout( $data, &$errors ) {
+			if ( 'kco' !== WC()->session->get( 'chosen_payment_method' ) ) {
+				return;
+			}
+
+			$klarna_order_id = WC()->session->get( 'kco_wc_order_id', 'missing' );
+			$klarna_order    = KCO_WC()->api->get_klarna_order( $klarna_order_id );
+			if ( is_wp_error( $klarna_order ) ) {
+				KCO_Logger::log( "[CHECKOUT VALIDATION]: Error getting Klarna order: {$klarna_order->get_error_message()}. For order ID: '$klarna_order_id'. Will not proceed with order." );
+				$errors->add( 'klarna_order', __( 'The Klarna order could not be retrieved from the session. Please try again.', 'klarna-checkout-for-woocommerce' ) );
+				return;
+			}
+
+			$billing_address = array_filter(
+				$data,
+				function ( $field ) {
+					return strpos( $field, 'billing_' ) === 0;
+				},
+				ARRAY_FILTER_USE_KEY
+			);
+
+			$shipping_address = array_filter(
+				$data,
+				function ( $field ) {
+					return strpos( $field, 'shipping' ) === 0;
+				},
+				ARRAY_FILTER_USE_KEY
+			);
+
+			// Retrieve the address from Woo.
+			$address_fields = array( 'first_name', 'last_name', 'company', 'address_1', 'address_2', 'city', 'state', 'postcode', 'country' );
```

Comment:
Just a suggestion, change this to be an array like:
```php
$address_fields = array(
    'first_name' => 'given_name',
    ...
)
```

Then you can avoid the switch later on each field and simply do `$wc_name = $address_fields[ $field ] ?? '';` in the loop and avoid having to do a switch case lookup each time.